src/main/java/de/tum/in/www1/artemis/config/migration/entries/BambooMigrationService.java

```diff
@@ -0,0 +1,499 @@
+
+        var response = restTemplate.exchange(builder.build().toUri(), HttpMethod.GET, null, String.class);
+        return getDataItemIds(response.getBody());
+    }
+
+    /**
+     * Returns the id of the repository with the given name for the given build plan.
+     * If no repository with the given name exists, an empty optional is returned.
+     *
+     * @param buildPlanId The key of the build plan, e.g. 'EIST16W1-BASE'.
+     * @param name        The name of the repository, e.g. 'assignment'.
+     * @return the id of the repository with the given name for the given build plan
+     */
+    private Optional<Long> getConnectedRepositoryId(String buildPlanId, String name) {
+        if ((name == null || name.isEmpty()) || (buildPlanId == null || buildPlanId.isEmpty())) {
+            return Optional.empty();
+        }
+        MultiValueMap<String, String> parameters = new LinkedMultiValueMap<>();
+        parameters.add("buildKey", buildPlanId);
+        String requestUrl = bambooServerUrl + "/chain/admin/config/editChainRepository.action";
+        UriComponentsBuilder builder = UriComponentsBuilder.fromUriString(requestUrl).queryParams(parameters);
+
+        var response = restTemplate.exchange(builder.build().toUri(), HttpMethod.GET, null, String.class);
+        var html = response.getBody();
+        var ids = getDataItemIds(html);
+        for (var id : ids) {
+            var repositoryName = getRepositoryNameById(html, id);
+            // old repositories have the name "Assignment", new ones "assignment"
+            if (repositoryName.isPresent() && repositoryName.get().equalsIgnoreCase(name)) {
+                return Optional.of(id);
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public void overrideRepositoriesToCheckout(String buildPlanKey, List<AuxiliaryRepository> auxiliaryRepositoryList) {
+        Optional<Long> testRepositoryId = getConnectedRepositoryId(buildPlanKey, "tests");
+        Optional<Long> assignmentRepositoryId = getConnectedRepositoryId(buildPlanKey, "assignment");
+
+        if (testRepositoryId.isEmpty()) {
+            log.error("Repository tests not found for build plan {}", buildPlanKey);
+        }
+        if (assignmentRepositoryId.isEmpty()) {
+            log.error("Repository assignment not found for build plan {}", buildPlanKey);
+        }
+
+        Map<String, Long> auxiliaryRepositoryIds = new HashMap<>();
+        for (AuxiliaryRepository auxiliaryRepository : auxiliaryRepositoryList) {
+            Optional<Long> repositoryId = getConnectedRepositoryId(buildPlanKey, auxiliaryRepository.getName());
+            if (repositoryId.isPresent()) {
+                auxiliaryRepositoryIds.put(auxiliaryRepository.getName(), repositoryId.get());
+            }
+            else {
+                log.error("Repository {} not found for build plan {}", auxiliaryRepository.getName(), buildPlanKey);
```

Comment:
If we end up here, the old linked repositories will already have been deleted and replaced with the new git repositories. If we can not find a repository by that name, because bamboo is not responding, then we now throw an Exception, thanks for the suggestion! Rolling back is not possible in this stage, if it fails here, all repositories that are checked out in this build plan will default to the first repository set in the configuration of the build plan. 

In case this happens, I doubt it, we can rerun the migration, the migration can be done multiple times and will fix the previous errors, each repository URL, name, etc. is saved in the Artemis database and is not fetched from the CI or VCS.

Reply:
Ok, sounds great!